### ISSUE TITLE
Stochastic computation graphs basic API

### ISSUE THREAD

**Issue by HEmile** [HEmile/storchastic#1]
Create the stochastic computation graph API.  

## API overview
Storchastic contains storchastic specific tensors `storch.Tensor`, that wrap `torch.Tensor` with a boolean containing whether it is a stochastic or deterministic node, and optionally the gradient estimation method.

`storch.sample` takes a sample method like `Reparameterization`, `Score` or `Expect` and returns a stochastic `storch.Tensor` with the associated sample method. A `storch.Tensor` cannot be used like other `torch.Tensor`s: They first have to be unwrapped through other `storch` functions to ensure they are properly registered to the graph. Whenever `storch.sample` is called, a stochastic node is created.

Whenever we want to use `storch.Tensor`s, whether deterministic or stochastic, they have to be called through a higher-order `storch` function, `storch.deterministic`, `storch.stochastic` or `storch.cost`. These receive a function and a list of inputs for this function, and applies the inputs to the function. This unwraps any `storch.Tensor` back into its `torch.Tensor` format so it can be used within regular PyTorch code. Furthermore, it creates appropriate dependencies between the inputs and outputs. 

`storch.deterministic` creates a context in which all behavior is deterministic, that is, it does not contain `storch.sample` calls. If `storch.sample` is called within a deterministic context, an exception is thrown. The outputs of a `storch.deterministic` call are deterministic `storch.Tensor`, unless the `storch.deterministic` call is within a deterministic context. In this case a normal `torch.Tensor`is returned.

`storch.stochastic` creates a context in which stochastic behavior is allowed. This creates dependencies between the functions inputs and any sample step. Note that `storch.stochastic` cannot be called from within a stochastic context: this would possibly create unclear dependencies. Furthermore, the function used in `storch.stochastic` should always return stochastic `storch.Tensor`s.

In addition, `storch.cost` is a deterministic context that requires that the output of the function is a single floating point tensor. This cost is registered to the storchastic graph. By calling `storch.backward()`, all registered costs are used to create a loss function, that includes the appropriate score functions where necessary, in addition to variance reduction methods. 

### Storch Tensor
`class storch.Tensor`. One is not supposed to create or interface with `storch.Tensor`. Storchastic automatically manages its creation and unwrapping. 

`_from_tensor(tensor, stochastic, method=None, cost=False)` wraps the tensor `tensor`, alongside a boolean whether it is a stochastic or deterministic node. If it is a stochastic node, one should also provide a `storch.EstimationMethod`. 

`_tensor()` returns the wrapped tensor. Only use this for debugging, as using the Tensor will break the assumptions of the graph.

### Storch sample
- [x] `sample(distribution, method=None, N=1)-> storch.Tensor` Creates a stochastic `storch.Tensor` with the given `method`. 
- [x] If `method` is not provided, it chooses between `Reparameterization` and `Score` depending on the computation graph for the continuous distribution and `Score` for discrete distributions. See issue #14. 
- [x] `N` refers to the amount of samples taken for the gradient estimation method. See issue #15.

### Storch deterministic
- [x] `deterministic(fn, inputs) -> [storch.Tensor] | fn:[]->[(torch.Tensor, storch.Tensor)]` Applies `fn` to the `inputs`. `fn` should return one or multiple `torch.Tensor`s or a `storch.Tensor`s. `inputs` can include `storch.Tensor`'s, which `deterministic` will unwrap and register to the graph as dependencies to its output `storch.Tensor`. 
- [x] `fn` should not call `storch.sample` or `storch.stochastic`.

### Storch stochastic
- [x] `stochastic(fn, inputs) -> [storch.Tensor] | fn:[]->[storch.Tensor]` Applies `fn` to the `inputs`. `fn` should return one or multiple `storch.Tensor`s. `inputs` can include `storch.Tensor`s. 
- [x] `fn` should not call `storch.stochastic`. 

### Storch cost
- [x] `cost(fn, inputs) -> [storch.Tensor]  | fn:[]->[(torch.Tensor, storch.Tensor)]` Like `deterministic`. 
- [x] However, `fn` should always return a single floating point tensor, that represents the cost tensor. This takes into account the size of the sample taken. 
- [x] In addition to registering the output of `fn` as a deterministic node to the graph, it adds that this node is a cost node and that its value should be minimized. 

### Decorators
- [x] To allow for easier and cleaner code, we provide decorators for `cost`, `stochastic` and `deterministic`. Specifically:
```
@storch.cost
def f(x):
    return torch.mean((mu-x)**2)
```

We can call `f(x)` using both `torch.Tensor` and with `storch.Tensor`. If called with a `storch.Tensor`, it unwraps the `storch.Tensor` automatically without having to call a storch specific unwrapping function. It also automatically registers it as a cost function.
